### ISSUE TITLE
Improving working with multiple selected records in Table

### ISSUE THREAD

**Issue by chris-fox** [Esri/solutions-components#321]
This is mostly in response to a user experience issue. Because the table supports multiselect, it makes it more challenging to work with one record at a time. For example, if I select one record and make an edit and then want to select the next record and edit it, I either need to unselect the first record or potentially use the Next navigation.

To this point, I noticed that in the Information Panel if you have more than 1 record selected it looks to order them in ascending order by ObjectID. So for example if I select features with OIDs 1, 5, and 3. They are ordered 1,3,5 no matter which order I selected them in the table.

So for example lets say I select OID 1, make an edit and then select OID 3 to make an edit. The information panel refreshes and is set to record OID 1 and then I have to click next to edit OID 3.

If I do the opposite, select OID 3, make an edit and the select OID 1 to make an edit. The information panel refreshes and is set to OID 1, so it is on the record I selected last and I can just edit it.

So my ideas are (not sure what is possible:

1. Persist the order you select records in. So they are ordered from most recently selected record to oldest selected record. That way When I am item 1 of X, 1 is always the record I most recently selected in Table or Map.
2. Keep the same order we have today, but set the index of the Information panel to the record that was most recently selected. So in the example above where I select OID 1 and then OID 3, it is sorted in the panel as 1,3, but the information panel display at record 2 of 2 so I am viewing and can edit the most recently selected record.

Personally, I prefer option #1. Thoughts?

**Comment by avezina** [comment]
I'm not sure. An additional case is when I select both from the Table and from the Map. When selecting from the Map, there's a chance I will select multiple features at once. What's expected in that case, especially if I already had records selected from the Table?

**Comment by azizaparveen** [comment]
@chris-fox 
I like the 2nd option better, in your example, if I already select and edit OID 1 and then select OID 3, I think it is a better experience to show the most recent selected feature (OID 3) in the panel and I can just edit it.
to Alix's point when I select three records at the same time and like to edit all of them one by one, I am ok with either ascending or descending order.

**Comment by avezina** [comment]
@chris-fox @azizaparveen

I looked at how Experience Builder handles this with the new [add to selection](https://devtopia.esri.com/Beijing-R-D-Center/ExperienceBuilder-Web-Extensions/issues/7093) functionality in devext and the Feature Info widget.
It looks like they opted for somewhat a mix of both options you described above @chris-fox:

https://github.com/Esri/solutions-components/assets/50338557/62bd31fb-b536-4e5a-bc35-8e7971a55029


1. Persist in which order you selected the records
AND
2. Set the index of the Information panel to the record that was most recently selected, and preserve that index while adding to the selection. 
3. If the currently active/index record is removed from the selection, then fall back to the first record listed (e.g. 1 of 10)

That behavior seemed smooth enough to me, but I can see the benefit of always loading the last selected record in the information panel too, instead of always keeping the first selected record active.

Maybe the following would work for us - if we are able to achieve this. Here's how I'd outline all the selection logic:

1. If ONE record is selected for the first time, show that record in the info panel
2. If MULTIPLE records are selected for the first time, load the selected records in the Info panel and order them by OID with the smallest OID value showing first (e.g. show records in panel starting at 1 of 10 with smallest OID at 1 and largest OID at 10)
3. When ADDING TO an existing selection
    - Preserve in which order you select the additional record or records
        - If only ONE additional record is selected by clicking on the map or in the table, then add this record to the info panel in sequence and show that record as active. 
            - For example, if there were already 10 records selected, the new selected record becomes 11, even if the OID is smaller than the one for the already selected records. (e.g. OID5,6,...,9,15, + **2**)
        - If MULTIPLE records are added to the selection at once by clicking in the map, then add these records to the info panel in sequence but keep the previously active record indexed and displayed in the info panel. 
            - For example, if there were already 10 records selected, the new selected records are added to the list and can be added in sequence based on their OID. (e.g. 23,65,12,32,...,98,34, + **1,2,3,4,5**)
4. If the currently active/index record is removed from the selection, then fall back to the first record listed in the sequence(e.g. 1 of 10) and do not reorder records in the sequence.

**Comment by chris-fox** [comment]
@avezina, what you describe sounds good to me.

**Comment by avezina** [comment]
Added one case 4 above for remove from the selection.

**Comment by azizaparveen** [comment]
@jmhauck do you think we will have time to implement this in beta?

**Comment by avezina** [comment]
Adding this here: ![MicrosoftTeams-image](https://github.com/Esri/solutions-components/assets/50338557/c813030c-7e9d-460a-8a09-d0ec23dd6273)

It seems to me the JS team may need to also find a way to handle the workflows we currently have outlined above.
Should we wait to understand what they are building, or perhaps we should talk to the team to see if what we're planning to build for our workflows would be useful to them?
@chris-fox any thoughts?

**Comment by chris-fox** [comment]
@avezina, yes I think it would be worth understanding what this will provide. However, I do feel in the short term we should see what we can do today to improve this experience. Most importantly for me is when I select a new record from a single check within the table I want that item to be the current item you are viewing in the information panel.

**Comment by avezina** [comment]
@chris-fox 
Discussed [Selection Manager](https://devtopia.esri.com/WebGIS/arcgis-js-api/issues/55338) with Craig. 

> Takeaways:
> - It's not really going to help us with selection logic itself. Any selection logic meant to represent things in a certain sequence in the Information panel is on us.
> - We'll probably still want to go back in the future and swap some of our logic to leverage the core Selection Manager view model they'll build. They are going to add some interesting highlight functionality we may be interested in leveraging.

Also talked about the challenges of handling our selection workflow. I'm aware that the rules I've outlined [here](https://github.com/Esri/solutions-components/issues/321#issuecomment-1719840356) are (maybe too?) complex, and perhaps we can think of a more simple and consistent behavior?

Did some more research and testing in other products, and talked to @gelisajoseph who's coming from an outside perspective, and came to the conclusion that:
- the "additive" behavior when clicking on the map may be increasing the complexity of our workflow, while also being a bit confusing if also selecting in the table.
- Another tricky requirement is always having that last selected record be the active one in the Info panel. 

Here's what we could do (and I think this is closer to what you had originally suggested @chris-fox):
- Selecting records in the **Table** is additive
  - newly selected records (check the box) are added to the selection set
  - the last selected record is shown in the Info panel (sequence based on user's selection)
- Selecting records from the **map** is _not_ additive
  - when selecting in the map, it always becomes a new selection set
  - this new selection set can be ordered based on OID when loaded
  - UX challenge 1:
    - when selecting records from the map, the selected records may not be visible in the table if the rows are not in the current scrolled view. A user needs to click "Show selected". I believe this button could be displayed in the main ribbon next to "Clear selection".
    ![image](https://github.com/Esri/solutions-components/assets/50338557/e72fb534-2e70-4dda-8bd0-ce84f26c1ca3)
  - UX challenge 2:
    - when selecting multiple records from the map (could also apply with table), a user will need to navigate through each popup in the Info panel until they find their record of interest. I come back with the suggestion of implementing the navigation list to help with this.
    ![image](https://github.com/Esri/solutions-components/assets/50338557/e2f742cb-8cb2-47d9-99a1-275b70162ede)
    ![image](https://github.com/Esri/solutions-components/assets/50338557/1a0d3d75-aae0-4052-b60c-90f142a08411)

- I think these rules can still apply:
  - If the current record is removed from the selection, then fall back to the first record listed in the sequence (e.g. 1 of 10) and do not reorder records in the sequence.
  - When switching selection, reorder everything based on OID, and fall back to the first record in the selection set array to be displayed in the Info panel.

Thoughts?

**Comment by chris-fox** [comment]
@avezina, I think this all makes sense. I am good with moving the Show Selected to the main toolbar and agree the navigation with multiple records will be a nice enhancement (Although not critical for Beta).

In the future I could see us leveraging the same Selection component in the Web Editor to support map selections. This would give the user control over whether map selections are additive or subtractive or a new selection completely. I also like the ability in the future to allow users to click and drag to select multiple records (Again this could be post beta)

For Beta I would like to do the following:

- Selecting records in the Table is additive
   - newly selected records (check the box) are added to the selection set
   - the last selected record is shown in the Info panel (sequence based on user's selection)
- Selecting records from the map is not additive
   - when selecting in the map, it always becomes a new selection set
   - this new selection set can be ordered based on OID when loaded
- If the current record is removed from the selection, then fall back to the first record listed in the sequence (e.g. 1 of 10) and do not reorder records in the sequence. **This is the current behavior**
- When switching selection, reorder everything based on OID, and fall back to the first record in the selection set array to be displayed in the Info panel. **This is the current behavior**

**Comment by avezina** [comment]
@chris-fox sounds good!

@jmhauck Please follow the list below for implementation purposes:

- [ ] Move the Show Selected to the main toolbar
- [ ] Selecting records in the Table is **additive** to the current selection set
  - newly selected records (check the box) are added to the selection set
  - the last selected record is shown in the Info panel (sequence based on user's selection)
- [ ] Selecting records from the map is **not additive**
  - when selecting in the map, it always becomes a new selection set
  - this new selection set can be ordered based on OID when loaded

Also, verify the behavior for:
- [ ] If the current record is removed from the selection, then fall back to the first record listed in the sequence (e.g. 1 of 10) and do not reorder records in the sequence.
- [ ] When switching selection, reorder everything based on OID, and fall back to the first record in the selection set array to be displayed in the Info panel.

**Comment by avezina** [comment]
@jmhauck Gave this a quick test, observing that records selected in the table are no longer showing as selected in the map:
![image](https://github.com/Esri/solutions-components/assets/50338557/cf5b068b-9155-4d43-953f-a902384e6ef0)

**Comment by azizaparveen** [comment]
verified in devext
